pkg/resource/cluster/hook.go

```diff
@@ -0,0 +1,278 @@
+	)
+	return ackrequeue.NeededAfter(
+		errors.New(msg),
+		RequeueAfterUpdateDuration,
+	)
+}
+
+// clusterHasTerminalStatus returns whether the supplied cluster is in a
+// terminal state
+func clusterHasTerminalStatus(r *resource) bool {
+	if r.ko.Status.Status == nil {
+		return false
+	}
+	cs := *r.ko.Status.Status
+	for _, s := range TerminalStatuses {
+		if cs == s {
+			return true
+		}
+	}
+	return false
+}
+
+// clusterActive returns true if the supplied cluster is in an active status
+func clusterActive(r *resource) bool {
+	if r.ko.Status.Status == nil {
+		return false
+	}
+	cs := *r.ko.Status.Status
+	return cs == StatusActive
+}
+
+// clusterCreating returns true if the supplied cluster is in the process of
+// being created
+func clusterCreating(r *resource) bool {
+	if r.ko.Status.Status == nil {
+		return false
+	}
+	cs := *r.ko.Status.Status
+	return cs == StatusCreating
+}
+
+// clusterDeleting returns true if the supplied cluster is in the process of
+// being deleted
+func clusterDeleting(r *resource) bool {
+	if r.ko.Status.Status == nil {
+		return false
+	}
+	cs := *r.ko.Status.Status
+	return cs == StatusDeleting
+}
+
+// returnClusterUpdating will set synced to false on the desired resource and
+// return an async requeue error to signify that the resource should be
+// forcefully requeued in order to pick up the 'UPDATING' status.
+func returnClusterUpdating(desired *resource, latest *resource) (*resource, error) {
+	msg := "Cluster is currently being updated"
+	ackcondition.SetSynced(desired, corev1.ConditionFalse, &msg, nil)
+	return desired, requeueAfterAsyncUpdate(latest)
+}
+
+func (rm *resourceManager) customUpdate(
+	ctx context.Context,
+	desired *resource,
+	latest *resource,
+	delta *ackcompare.Delta,
+) (updated *resource, err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.customUpdate")
+	defer exit(err)
+
+	if clusterDeleting(latest) {
+		msg := "Cluster is currently being deleted"
+		ackcondition.SetSynced(desired, corev1.ConditionFalse, &msg, nil)
+		return desired, requeueWaitWhileDeleting
+	}
+	if !clusterActive(latest) {
+		msg := "Cluster is in '" + *latest.ko.Status.Status + "' status"
+		ackcondition.SetSynced(desired, corev1.ConditionFalse, &msg, nil)
+		if clusterHasTerminalStatus(latest) {
+			ackcondition.SetTerminal(desired, corev1.ConditionTrue, &msg, nil)
+			return desired, nil
+		}
+		return desired, requeueWaitUntilCanModify(latest)
+	}
+
+	// Merge in the information we read from the API call above to the copy of
+	// the original Kubernetes object we passed to the function
+	ko := desired.ko.DeepCopy()
+
+	// None of these methods modify the status, so we should return the latest
+	// status as given by the ReadOne
+	ko.Status = latest.ko.Status
+
+	if delta.DifferentAt("Spec.Logging") {
+		if err := rm.updateConfigLogging(ctx, desired); err != nil {
+			awserr, ok := ackerr.AWSError(err)
+
+			// The API responds with an error if there were no changes applied
```

Comment:
le sigh... :)

Reply:
:) I think I'll need to add additional logic for the logging values. If we add an element that enables just one of the logging options, the describe call returns two elements - the enabled logging config (same as the spec) and another config which shows every other option as disabled. Our delta currently sees two slice elements and assumes its changed, calling update on every reconciliation loop.

Reply:
++

To be clear, I was "le sigh"ing the behaviour of the EKS API returning an error when no changes were made, not "le sigh"ing anything to do with your code! :)